qiskit_dynamics/solvers/dyson_magnus_solvers.py

```diff
@@ -0,0 +1,995 @@
+        self._dt = dt
+
+        # compute perturbative terms
+        results = solve_lmde_perturbation(
+            perturbations=perturbations,
+            t_span=[0, dt],
+            expansion_method=expansion_method,
+            expansion_order=expansion_order,
+            expansion_labels=expansion_labels,
+            integration_method=integration_method,
+            **kwargs,
+        )
+
+        if self.expansion_method == "dyson":
+            # if Dyson multiply by single step frame transformation
+            results.perturbation_results.expansion_terms = (
+                Array(self.Udt) @ results.perturbation_results.expansion_terms
+            )
+            new_coeffs = results.perturbation_results.expansion_terms[:, -1]
+            self._expansion_polynomial = ArrayPolynomial(
+                array_coefficients=new_coeffs,
+                monomial_labels=results.perturbation_results.expansion_labels,
+                constant_term=self.Udt,
+            )
+        elif self.expansion_method == "magnus":
+            self._expansion_polynomial = ArrayPolynomial(
+                array_coefficients=results.perturbation_results.expansion_terms[:, -1],
+                monomial_labels=results.perturbation_results.expansion_labels,
+            )
+
+    @property
+    def expansion_method(self):
+        """Perturbation method used in solver."""
+        return self._expansion_method
+
+    @property
+    def dt(self):
+        """Step size of solver."""
+        return self._dt
+
+    @property
+    def Udt(self):
+        """Single step frame transformation."""
+        return self._Udt
+
+    @property
+    def operators(self):
+        """Original operators in the generator."""
+        return self._operators
+
+    @property
+    def rotating_frame(self):
+        """Rotating frame."""
+        return self._rotating_frame
+
+    @property
+    def expansion_polynomial(self) -> ArrayPolynomial:
+        """ArrayPolynomial object for evaluating the perturbation series."""
+        return self._expansion_polynomial
+
+    def approximate_signals(self, signals: List[Signal], t0: float, n_steps: int) -> np.ndarray:
+        """Approximate a list of signals over a series of intervals according to the Chebyshev
+        time-step and Chebyshev approximation structure set at instantiation.
+
+        Args:
+            signals: List of Signals to approximate.
+            t0: Start time.
+            n_steps: Number of time steps to approximate over.
+
+        Returns:
+            np.ndarray: The Chebyshev coefficients of each signal over each time-step.
+            The first dimension indexes the signals, and the second the time-step.
+        """
+        return self._approximate_signals(signals, t0, n_steps)
+
+    def evaluate(self, coeffs: Array) -> Array:
+        """Evaluate the expansion polynomial for a given list of coefficients."""
+        return self.expansion_polynomial(coeffs)
+
+
+def perturbative_solve(
```

Comment:
If these are all intended as private methods, might be good to add an underscore just to make this clear

Reply:
Ended up deciding to go with the underscores - changed every class/method that is internal to have an underscore at the beginning of the name.